[LoongArch64] add Intrinsics' API for LoongArch64.

shushanhf opened the pull request:
We have finished the SIMD on the runtime6.0 and the tests passed.

I will push the SIMD for LoongArch64.

This is the first PR about the API's name.

The  [[API Proposal]: LoongArch64: add Intrinsics' API for LoongArch64](https://github.com/dotnet/runtime/issues/94445#top)
#94445

@tannergooding
Can you give me some advices ?
Thanks

shushanhf committed:
[LoongArch64] add Intrinsics' API for LoongArch64.

shushanhf commented:
@tannergooding
Can you give me some advices ?

This is just the API's name, and first focus on the class name and the API name.

Later I will update this PR to amend some details.

Thanks

huoyaoyuan commented:
As a new architecture, it's more risky to expose public APIs comparing to mature architectures. I'd suggest keeping them internal, and focusing on cross-platform Vector128/256 intrinsics now.

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch64/LA64Base.PlatformNotSupported.cs:
```diff
@@ -0,0 +1,217 @@
+#else
+    internal
+#endif
+    abstract class LA64Base
```
The name `LA64Base` is about some instructions within the LoongArch64's general ISA.

huoyaoyuan commented:
For API names, you can open API proposal like #94011. API definition without JIT implementation should be unwanted.

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch64/LA_LASX.cs:
```diff
@@ -0,0 +1,2269 @@
+    /// </summary>
+    [Intrinsic]
+    [CLSCompliant(false)]
+    public abstract class LA_LASX : LA_LSX
```
The `LSX` is the abbreviation of Loongson SIMD Extension which is SIMD-128bits.
The `LASX` is the abbreviation of Loongson Advanced SIMD Extension which is SIMD-256bits.

tannergooding replied:
We likely would just call these `Lasx` and `Lsx`, to match the normal .NET naming convention and roughly what we've done for other platforms.

shushanhf replied:
OK, I will update these later.
Thanks.

shushanhf commented:
> For API names, you can open API proposal like #94011. API definition without JIT implementation should be unwanted.

If the API is OK for LoongArch64, I will push the JIT implementation.

shushanhf commented:
> As a new architecture, it's more risky to expose public APIs comparing to mature architectures. I'd suggest keeping them internal, and focusing on cross-platform Vector128/256 intrinsics now.

yes, the  Vector128/256  is independent of the CPU.

Now the API for architecture is the most important for LoongArch64, I want to confirm them for LoongArch64.
And then I will push all the Hardware-Intrinsics and SIMD.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch64/LA64Base.PlatformNotSupported.cs:
```diff
@@ -0,0 +1,217 @@
+#pragma warning disable IDE0060 // unused parameters
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.Intrinsics.LoongArch64
```
I've marked this as `NO-MERGE` since we cannot take it until after an API review has occurred. See https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md

We need an API proposal, following the standard template, created first. We'll have the discussion on relevant name changes and other bits there, then I can then champion that and take it to API review. Once approved, we can then implement the API surface.

Until then, LoongArch would be relegated to only supporting the existing cross platform API surface. For example, `Leading/TrailingZeroCount` can be supported by accelerating `int.Leading/TrailingZeroCount` and the same methods on the other primitive types.

shushanhf replied:
Thanks !
Reviewing the API for LoongArch64 based on a PR maybe more clear. So I pushed this PR.
I will create an API proposal for LoongArch64's API.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch64/LA_LASX.cs:
```diff
@@ -0,0 +1,2269 @@
+        /// float64x4_t xvfmin_d_f64 (float64x4_t a, float64x4_t b)
+        ///   LASX: XVFMIN.D Xd.4D, Xj.4D, Xk.4D
+        /// </summary>
+        public static Vector256<double> Min(Vector256<double> left, Vector256<double> right) => Min(left, right);
```
What's the semantics around `NaN` and `-0` handling on LoongArch?

shushanhf replied:
The float operation is implemented within the IEEE-754-2008, here is MinNum(x,y).

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch64/LA64Base.cs:
```diff
@@ -0,0 +1,212 @@
+
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.Intrinsics.LoongArch64
```
Is there any `LoongArch32` that needs to be considered? If so, does it essentially have the same set of SIMD instructions, etc?

For Arm and x86, we have a single namespace encompassing both the 32-bit and 64-bit processors. The unique ISAs are then exposed as `Sse.X64.Method()` or `AdvSimd.Arm32.Method()` and `AdvSimd.Arm64.Method()`

shushanhf replied:
> Is there any `LoongArch32` that needs to be considered? If so, does it essentially have the same set of SIMD instructions, etc?

If just considering the current CoreCLR's implementation which only supports the LoongArch64, of course the SIMD extension is only  binding with the 64-bits CPU.
Here using `LoongArch64` directly just for there is no a LoongArch32's implementation CoreCLR to test the SIMD.

I will update it as `namespace System.Runtime.Intrinsics.LoongArch` liking Arm and x86.
Thanks.

shushanhf committed:
Rename LA64Base -> LoongArchBase, LA_LSX -> Lsx, LA_LASX -> Lasx.
Update the API within the LoongArchBase class.

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch64/LA_LASX.cs:
```diff
@@ -0,0 +1,2269 @@
+        /// bool xvsetnez_v_u8 (uint8x32_t value)
+        ///   LASX: XVSETNEZ.V cd, Xj.32B
+        /// </summary>
+        public static bool HasElementsNotZero(Vector256<byte> value) => HasElementsNotZero(value);
```
I will answer these together later.

shushanhf replied:
> How does this instruction work at the hardware level?
> 
> `Xj.32B` is clearly the input register, but I'm not familiar with `cd` here. Is it a general purpose register, a flag register, something else?

The `cd` is a float flag register which indicating the floats comparing results.
There are 8 `cd` float flag registers.

Of course here I didn't expose the `cd` the API just for simple usage.

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch/LoongArchBase.cs:
```diff
@@ -0,0 +1,198 @@
+    /// </summary>
+    [Intrinsic]
+    [CLSCompliant(false)]
+    public abstract class LoongArchBase
```
* Rename this file as `LoongArchBase.cs`, is it OK? 
Or Just name this file as `LABase.cs`?
* Naming this class as `LoongArchBase`, is it OK ?

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch/LoongArchBase.cs:
```diff
@@ -0,0 +1,198 @@
+            /// <summary>
+            ///   LA64: CLO.W rd, rj
+            /// </summary>
+            public static int LeadingSignCount(int value) => LeadingSignCount(value);
+
+            /// <summary>
+            ///   LA64: CLO.W rd, rj
+            /// </summary>
+            public static int LeadingSignCount(uint value) => LeadingSignCount(value);
```
Is it needed to add two types API with `int value` and `uint value` ?

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch/LoongArchBase.cs:
```diff
@@ -0,0 +1,198 @@
+            /// <summary>
+            ///   LA64: BITREV.W rd, rj
+            /// </summary>
+            public static long ReverseElementBits(int value) => ReverseElementBits(value);
+
+            /// <summary>
+            ///   LA64: BITREV.W rd, rj
+            /// </summary>
+            public static ulong ReverseElementBits(uint value) => ReverseElementBits(value);
```
Is it needed to add the `int value` and `uint value` for the API `ReverseElementBits()` ?

shushanhf reviewed (commented):
@tannergooding

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch/LoongArchBase.cs:
```diff
@@ -0,0 +1,198 @@
+            /// <summary>
+            ///   LA64: REVB.2W rd, rj
+            /// </summary>
+            public static int ReverseElementBits(int value) => ReverseElementBits(value);
+
+            /// <summary>
+            ///   LA64: REVB.2W rd, rj
+            /// </summary>
+            public static uint ReverseElementBits(uint value) => ReverseElementBits(value);
+
+            /// <summary>
+            ///   LA64: REVB.D rd, rj
+            /// </summary>
+            public static long ReverseElementBits(long value) => ReverseElementBits(value);
+
+            /// <summary>
+            ///   LA64: REVB.D rd, rj
+            /// </summary>
+            public static ulong ReverseElementBits(ulong value) => ReverseElementBits(value);
```
These are part of instructions liking the Arm64's `REV, REV16, REV32, REV64`, but the `ArmBase class` doesn't support these, Why?

Is it needed to add these for LoongArch64.

shushanhf committed:
Lsx-PartA: add some APIs for Lsx.

shushanhf committed:
Lsx-PartB, Lasx-PartA: add some bitwise ops, MultiplyHight, Mod,
    Sign-Zero-extend and MultiplyWiden

shushanhf committed:
Lsx-PartC, Lasx-PartB: some Shifting operations with narrowing and rounding.
Count leading ones/zeros and elements' bit clear.

shushanhf committed:
Lsx-PartD, Lasx-PartC: amend the code formate.
Add more ADD's operations.

shushanhf committed:
Lsx-PartE, Lasx-PartD: add some add/madd operations, max/min,
bitwise shift, shuffle, compare and float operations.

shushanhf committed:
Lsx-PartF, Lasx-PartE: amend the Load/Store Vector128/256,
    add LoadElementReplicateVector, Vector elements' operations and AverageRounded.

shushanhf commented on src/libraries/System.Private.CoreLib/src/System/Runtime/Intrinsics/LoongArch64/LA_LASX.cs:
```diff
@@ -0,0 +1,2269 @@
+        /// float32x8_t xvfrecip_s_f32 (float32x8_t a)
+        ///   LASX: XVFRECIP.S Xd.8S Xj.8S
+        /// </summary>
+        public static Vector256<float> Reciprocal(Vector256<float> value) => Reciprocal(value);
```
The `Reciprocal` is implemented with the IEEE754-2008   division(1.0,x).

shushanhf replied:
Only the `FRECIPE` and `FRSQRTE` within the `LoongArchBase` class are extimate.
But the  `FRECIP` and `FRSQRT` are exact.

shushanhf committed:
LoongArchBase: add CRC-32 and some float operations.

shushanhf committed:
update the *PlatformNotSupported.cs and ref/System.Runtime.Intrinsics.cs.

shushanhf commented:
Hi, @tannergooding 
I finished the first version of this PR, could you please review this PR?
Thanks

shushanhf commented:
> > could you please review this PR?
> 
> I can potentially give it a pass today or tomorrow, but its still blocked until API review can happen. That probably won't happen until the new year as API review typically doesn't happen in December when most people are on holiday/vacation.

OK, Thanks
I can wait more reviewers.

I will push other PRs that are independent of these APIs liking the SIMD's instructions within the emitter #95456

shushanhf committed:
add VectorSaturate and VectorSaturateUnsigned.
Also amend some code-formate.

tannergooding commented:
I'm still waiting for response to the question asked on the API proposal:

> Where is the spec for the LA64 SIMD ISA?
>
> Based on https://loongson.github.io/LoongArch-Documentation/README-EN.html, it looks like it should be Volume 2, but there doesn't appear to be any version published for it yet and the backing GitHub repo looks to be archived now and I do not see a replacement.

tannergooding commented:
Thanks! This is still on my backlog but is lower priority than some other work due to the API review not having happened yet (and this PR being blocked until that can happen).

I'll try to set some time aside in the next week or two to go through the SIMD ISA guide and compare it to the proposed API surface so that it can get marked `ready-for-review`

shushanhf commented:
> Thanks! This is still on my backlog but is lower priority than some other work due to the API review not having happened yet (and this PR being blocked until that can happen).
> 
> I'll try to set some time aside in the next week or two to go through the SIMD ISA guide and compare it to the proposed API surface so that it can get marked `ready-for-review`

OK, Thanks very much.
I will do it after the Feb. 17 for the Chinese Spring Festival.

steveisok commented:
@tannergooding what is the status of this review?

tannergooding commented:
As per the above, new public API surface cannot be added without it first going through API Review.

Last checked, the LoongArch specs weren't available or were still in draft form, so going to API review wasn't possible.

If that's changed, then it'd be great to have links and references to the official docs so I can help drive this through to completion.
